File: cpp/116_PopulatingNextRightPointersinEachNode.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if(!root) return;
        queue<TreeLinkNode *> q;
        q.push(root);
        q.push(NULL);
        while(q.size()){
            auto t = q.front();
            q.pop();
            if(t != NULL){
                t->next = q.front();
                if(t->left) q.push(t->left);
                if(t->right) q.push(t->right);
            }else{
                if(q.size()) q.push(NULL);
            }
        }
    }
};

File: js/116_PopulatingNextRightPointersinEachNode.js
/**
 * Definition for binary tree with next pointer.
 * function TreeLinkNode(val) {
 *     this.val = val;
 *     this.left = this.right = this.next = null;
 * }
 */

/**
 * @param {TreeLinkNode} root
 * @return {void} Do not return anything, modify tree in-place instead.
 */
var connect = function(root) {
    if(!root) return;
    var queue = [];
    queue.push(root);
    queue.push(null);
    while(queue.length){
        var t = queue.shift();
        if(t !== null){
            t.next = queue[0];
            if(t.left) queue.push(t.left);
            if(t.right) queue.push(t.right);
        }else{
            if(queue.length) queue.push(null);
        }
    }
};